<a href="https://colab.research.google.com/github/Pranav-22-aditya/AI-WING-MODELS/blob/main/advertisement_targeting_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Targeting people for advertisement.In this model we will target the people who are most likely to buy a car .I have used the decision tree algorithm to find out which people are most likely to buy a car . this can be used for the advertisement of a car company in real life.

In [9]:
import pandas as pd


In [10]:
df = pd.read_csv('/content/drive/MyDrive/ai wing datasets/Advertisement_target_public.csv')

In [11]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [12]:
df

age        wc      education  ...   gender hours per week IncomeClass
0       38   Private        HS-grad  ...     Male             40       <=50K
1       28   Private      Bachelors  ...   Female             40       <=50K
2       37   Private        Masters  ...   Female             40       <=50K
3       31   Private        Masters  ...   Female             50        >50K
4       42   Private      Bachelors  ...     Male             40        >50K
...    ...       ...            ...  ...      ...            ...         ...
19782   53   Private        Masters  ...     Male             40        >50K
19783   22   Private   Some-college  ...     Male             40       <=50K
19784   40   Private        HS-grad  ...     Male             40        >50K
19785   58   Private        HS-grad  ...   Female             40       <=50K
19786   22   Private        HS-grad  ...     Male             20       <=50K

[19787 rows x 8 columns]

In [ ]:
df.isnull().sum()

age               0
wc                0
education         0
marital status    0
race              0
gender            0
hours per week    0
IncomeClass       0
dtype: int64

thus data wrangling is not required in this case as there is no missing data.however i am not a racist thus i am dropping the race column and as it doesn't effect our decision much.xD 

In [13]:
df.drop(columns='race', inplace=True)

In [ ]:
df

age        wc      education  ...   gender hours per week  IncomeClass
0       38   Private        HS-grad  ...     Male             40        <=50K
1       28   Private      Bachelors  ...   Female             40        <=50K
2       37   Private        Masters  ...   Female             40        <=50K
3       31   Private        Masters  ...   Female             50         >50K
4       42   Private      Bachelors  ...     Male             40         >50K
...    ...       ...            ...  ...      ...            ...          ...
19782   53   Private        Masters  ...     Male             40         >50K
19783   22   Private   Some-college  ...     Male             40        <=50K
19784   40   Private        HS-grad  ...     Male             40         >50K
19785   58   Private        HS-grad  ...   Female             40        <=50K
19786   22   Private        HS-grad  ...     Male             20        <=50K

[19787 rows x 7 columns]

In [14]:
df = pd.get_dummies(df, drop_first=True)

In [15]:
df

age  hours per week  ...  gender_ Male  IncomeClass_ >50K
0       38              40  ...             1                  0
1       28              40  ...             0                  0
2       37              40  ...             0                  0
3       31              50  ...             0                  1
4       42              40  ...             1                  1
...    ...             ...  ...           ...                ...
19782   53              40  ...             1                  1
19783   22              40  ...             1                  0
19784   40              40  ...             1                  1
19785   58              40  ...             0                  0
19786   22              20  ...             1                  0

[19787 rows x 16 columns]

In [17]:
x = df.iloc[:,:-1]
y = df.iloc[:, -1]

In [18]:
x

age  hours per week  ...  marital status_Married  gender_ Male
0       38              40  ...                       0             1
1       28              40  ...                       1             0
2       37              40  ...                       1             0
3       31              50  ...                       0             0
4       42              40  ...                       1             1
...    ...             ...  ...                     ...           ...
19782   53              40  ...                       1             1
19783   22              40  ...                       0             1
19784   40              40  ...                       1             1
19785   58              40  ...                       0             0
19786   22              20  ...                       0             1

[19787 rows x 15 columns]

In [19]:
y

0        0
1        0
2        0
3        1
4        1
        ..
19782    1
19783    0
19784    1
19785    0
19786    0
Name: IncomeClass_ >50K, Length: 19787, dtype: uint8

In [20]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [21]:
from sklearn.tree import DecisionTreeClassifier

In [22]:
dtc = DecisionTreeClassifier(random_state=0)

In [23]:
dtc.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [28]:
y_pred = dtc.predict(x_test)

In [29]:
from sklearn.metrics import confusion_matrix

In [30]:
cm = confusion_matrix(y_test, y_pred)
score = dtc.score(x_test, y_test)

In [31]:
print(cm)
print(score)

[[2539  339]
 [ 525  555]]
0.7817079332996463


thus the current accuracy of the model is 78.17%

Ensemble technique
1. Bagging with RandomForest

In [32]:
from sklearn.ensemble import RandomForestClassifier

In [33]:
rfc = RandomForestClassifier(random_state=0)

In [34]:
rfc.fit(x_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [35]:
cm2 = confusion_matrix(y_test, y_pred)
score2 = rfc.score(x_test, y_test)

In [36]:
#print(cm)
print(cm2)
#print(score)
print(score2)

[[2539  339]
 [ 525  555]]
0.7925720060636685


thus the accuracy of our model after enhacement by bagging with random forest method  is 79.25%.